SERPENT Interface

jbae11 opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)

#611 

##### What are the significant changes in functionality due to this change request?
RAVEN can now run SERPENT and stores its depletion results and criticality calculations

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

PaulTalbot-INL commented:
Also, check that there's no trailing whitespace in your new Python files; there's a script to do this in raven/developer_tools called "delete_trailing_whitespace.sh".  I believe you just give it the filename you want it to act on.

PaulTalbot-INL commented:
Hi, Jin,

Did you push the underscores changes for SerpentInterface.py?  It still shows with underscore variables for me.

Also, for doc strings, there should be a two-space indent after opening docstrings until closing them; for several of your new methods, the docstrings are flush with the quotation marks.  For example, this
```python
def myMethod(self,val):
  """
  @ In, val, float, value
  @ Out, None
  """
  pass
```
should be
```python
def myMethod(self,val):
  """
    @ In, val, float, value
    @ Out, None
  """
  pass
```

Lastly, the other review comments still need to be addressed.  @alfoa should be back in the next couple days and explain how to set up the interface test, as well.

alfoa commented:
Hi @jbae11 . Thank you very much for this contribution.

Tomorrow, I will create a new page in the Wiki that will explain how to create a test for RAVEN without the need to run the executable. I will post here once it is done.

Thanks.
Andrea

alfoa commented:
@jbae11 we created a page that explains how to create a test. You can find it in [here](https://github.com/idaholab/raven/wiki/contributing). Please let us know if it is not clear enough.
Thanks a lot.

jbae11 commented:
@alfoa  Thank you! I will get working on it!

alfoa commented:
@jbae11  I added a better description about how the ```tests``` file should look like. Look for the sentence " Add the output files of the driven code in the right directory ".
Just some hints:
* Your tests file must always have the tag ```type=="RavenFramework"``` 
* The ```input``` must be set to a RAVEN input that tests your interface (I say your RAVEN input file that you added in the tests will not work since no <Output> has been inputted in the <Step>

Your tests file (taking the info from your fork) should look like something like this:

```
[Tests]
 [./SerpentTest1]
  type = 'RavenFramework'
  input = 'aRealRAVENinputFile.xml'
  text = 'serpentTest1/sampleInput/1/testInput.serpent serpentTest1/sampleInput/1/testInput.serpent'
  # %serpent_output_root% must be replaced with the file root your interface useses to create the CSV RAVEN reads (e.g. my_serpent.csv)
  output = 'serpentTest1/sampleInput/1/%serpent_output_root%.csv'     
  # %RAVENoutstreamsPrintName% must be replaced with the name of the RAVEN OutStreams (of type Print) that you use to dump in CSV a RAVEN DataObject
  csv = 'serpentTest1/%RAVENoutstreamsPrintName%.csv' 
  test_interface_only = True 
 [../]
[]
```

alfoa commented:
@jbae11 is it a bit clearer? Can we help you somehow?

jbae11 commented:
@alfoa  Thank you for taking your time to do this. I think I understand what it is, but the term `output` is a bit ambiguous in the sense that is it:
1. output of SERPENT? (.res, .bumat files)
2. output of RAVEN? (the `CodeInterface` output parser creates a single csv for one `run` among multirun)
By looking at your example, it seems like it is the latter,  but then how are we testing the interface capability of combining the code ouput data (.res, .bumat files) to RAVEN-friendly csv files?

I hope I am making sense. Thank you so much for your patience.

alfoa commented:
@jbae11 

In the ```output``` tag you should list the files your interface is creating. 
For example, if you get from SERPENT the files ```my_serpent.res``` and  ```my_serpent.bumat``` and your interface reads them to create the RAVEN-readable CSV file named ```my_serpent.csv```, the ```output``` tag should look like the following:

```output= 'myWorkingDirectory/MyStepName/1/my_serpent.csv myWorkingDirectory/MyStepName/2/my_serpent.csv'```

since we want to be sure that your interface is able to read the SERPENT output files and convert them into a CSV.


In the tag ```csv``` or ```UnorderedCsv``` you should list the output of RAVEN ```OutStreams``` of type ```Print```. For example, if you have two ```OutStreams```:
``` 
 <OutStreams>
    <Print name="first_outstream">
       <type>csv</type>
       <source>myPointSet</source>
   </Print>
    <Print name="second_outstream">
       <type>csv</type>
       <source>myHistorySet</source>
   </Print>
 </OutStreams>
```
the ```csv``` tag would look like the following:

```csv= 'myWorkingDirectory/first_outstream.csv myWorkingDirectory/second_outstream.csv'```

Is it clearer?

Please let me know.

Andrea





@alfoa Thank you for taking your time to do this. I think I understand what it is, but the term output is a bit ambiguous in the sense that is it:

    output of SERPENT? (.res, .bumat files)
    output of RAVEN? (the CodeInterface output parser creates a single csv for one run among multirun)
    By looking at your example, it seems like it is the latter, but then how are we testing the interface capability of combining the code ouput data (.res, .bumat files) to RAVEN-friendly csv files?

I hope I am making sense. Thank you so much for your patience.

jbae11 commented:
@alfoa  Thank you.  I apologize for my delayed responses. I have made changes to what I think is the right way to set up the tests. Please let me know if there are any fixes that needs to made. Thank you again.

alfoa commented on framework/CodeInterfaces/SERPENT/testInput.xml:
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" ?>
+<!-- this example simply varies SERPENT deptime along a distribution -->
```
Please remove this file.

alfoa commented on framework/CodeInterfaces/SERPENT/generateCustomVariable.py:
```diff
@@ -0,0 +1,36 @@
+
+file = './aux-input-files/isoFile'
```
what is this for?

alfoa commented on framework/CodeInterfaces/SERPENT/aux-input-files/targetIsotopesKeff.xml:
```diff
@@ -0,0 +1 @@
+<Group name="targetSpace">d1001,d1002,d1003,d2003,d2004,d20060,d3006,d3007,d30080,d30090,d40080,d4009,d40100,d40110,d5010,d5011,d50120,d60120,d60130,d60140,d60150,d70130,d7014,d7015,d70160,d70170,d8016,d8017,d80180,d80190,d90180,d9019,d90200,d100200,d100210,d250610,d250620,d260580,d260590,d260600,d260610,d260620,d260640,d260650,d260660,d270590,d270600,d270601,d270610,d270620,d270630,d270640,d2706 ...
```
this is in the wrong location (you should remove it)

alfoa commented on framework/CodeInterfaces/SERPENT/aux-input-files/targetIsotopes.xml:
```diff
@@ -0,0 +1 @@
+<Group name="targetSpace">d1001,d1002,d1003,d2003,d2004,d20060,d3006,d3007,d30080,d30090,d40080,d4009,d40100,d40110,d5010,d5011,d50120,d60120,d60130,d60140,d60150,d70130,d7014,d7015,d70160,d70170,d8016,d8017,d80180,d80190,d90180,d9019,d90200,d100200,d100210,d250610,d250620,d260580,d260590,d260600,d260610,d260620,d260640,d260650,d260660,d270590,d270600,d270601,d270610,d270620,d270630,d270640,d2706 ...
```
this is in the wrong location (you should remove it)

alfoa commented on framework/CodeInterfaces/SERPENT/aux-input-files/serpentInput.serpent:
```diff
@@ -0,0 +1,148 @@
+set title "MSBR, no Th,U-233 8.5E-11, after 400 days 7.75E-1"
```
this is in the wrong location (you should remove it)

alfoa commented on framework/CodeInterfaces/SERPENT/aux-input-files/isoFile:
```diff
@@ -0,0 +1,1158 @@
+1001
```
this is in the wrong location (you should remove it)

alfoa commented on framework/CodeInterfaces/SERPENT/aux-input-files/geomFile:
```diff
@@ -0,0 +1,52 @@
+% --- Surfaces (core pitch = 10.16 cm):
```
this is in the wrong location (you should remove it)

alfoa commented on framework/CodeInterfaces/SERPENT/aux-input-files/featureIsotopes.xml:
```diff
@@ -0,0 +1 @@
+<Group name="featureSpace">f1001,f1002,f1003,f2003,f2004,f20060,f3006,f3007,f30080,f30090,f40080,f4009,f40100,f40110,f5010,f5011,f50120,f60120,f60130,f60140,f60150,f70130,f7014,f7015,f70160,f70170,f8016,f8017,f80180,f80190,f90180,f9019,f90200,f100200,f100210,f250610,f250620,f260580,f260590,f260600,f260610,f260620,f260640,f260650,f260660,f270590,f270600,f270601,f270610,f270620,f270630,f270640,f270 ...
```
this is in the wrong location (you should remove it)

alfoa commented on framework/CodeInterfaces/SERPENT/aux-input-files/customVariable.xml:
```diff
@@ -0,0 +1,1158 @@
+<variable name="f1001"/> 
```
this is in the wrong location (you should remove it)

alfoa commented on framework/CodeInterfaces/SERPENT/SerpentInterface.py:
```diff
@@ -0,0 +1,191 @@
+      @ In, fargs, dict, optional, a dictionary containing the axuiliary input file variables the user can specify in the input (e.g. under the node < Code >< clargstype =0 input0arg =0 aux0extension =0 .aux0/ >< /Code >)
+      @ Out, returnCommand, tuple, tuple containing the generated command. returnCommand[0] is the command to run the code (string), returnCommand[1] is the name of the output root
+    """
+    if clargs==None:
```
```if clargs is None```

alfoa commented on framework/CodeInterfaces/SERPENT/SerpentInterface.py:
```diff
@@ -0,0 +1,191 @@
+      raise IOError('No input file was specified in clargs!')
+    #check for output either in clargs or fargs
+    #if len(fargs['output'])<1 and 'output' not in clargs.keys():
+    #  raise IOError('No output file was specified, either in clargs or fileargs!')
```
can you remove commented code?

alfoa commented on framework/CodeInterfaces/SERPENT/SerpentInterface.py:
```diff
@@ -0,0 +1,191 @@
+    todo += executable
+    index=None
+    #inputs
+    for flag,exts in clargs['input'].items():
```
This looks copied/pasted from the GenericCode Interface. Do you need all these features for running Serpent?

alfoa replied:
(I mean all the code  from ```def getFileWithExtension(fileList,ext)``` to ```return returnCommand```

alfoa commented on framework/CodeInterfaces/SERPENT/SerpentInterface.py:
```diff
@@ -0,0 +1,191 @@
+    parser = GenericParser.GenericParser(infiles)
+    parser.modifyInternalDictionary(**Kwargs)
+    parser.writeNewInput(currentInputFiles,origfiles)
+    return currentInputFiles
```
this is copied/pasted by the Generic Code interface...

Can't' you just inherit from ```GenericCode```?

alfoa commented on framework/CodeInterfaces/SERPENT/SerpentInterface.py:
```diff
@@ -0,0 +1,191 @@
+
+
+  def finalizeCodeOutput(self, command, output, workDir):
+    # filename would be 'input.serpent'
```
docstrings is missing here.

alfoa commented on framework/CodeInterfaces/SERPENT/SerpentInterface.py:
```diff
@@ -0,0 +1,191 @@
+    outbumatfile = os.path.join(workDir, filename+".bumat1")
+    # get the list of isotopes to track
+    scriptLoc = os.path.dirname(os.path.realpath(sys.argv[0]))
+    isofile = os.path.join(scriptLoc, 'CodeInterfaces/SERPENT/aux-input-files/isoFile')
```
can you explain to me this thing here? are you forcing the Serpent User to have the isoFile in ```CodeInterfaces/SERPENT/aux-input-files/```?? And if so, why?

jbae11 replied:
My intention was that the user can customize what isotopes to track by using the `isoFile`, and from the `isoFile` the user can generate the `feature_space` and the `target_space`, using the `generateCustomVariable.py`. Thinking back this sounds really buggy, if there is a work around this please let me know. Thanks!

alfoa commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+          raise ValueError()
+        else:
+          hit = True
+          continue
```
```continue``` is not needed

alfoa commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+      if line.split(' ')[0] == 'dep':
+        if line.split(' ')[1] != 'daystep':
+          print('Currently can only take daystep')
+          raise ValueError()
```
```
print('Currently can only take daystep')
raise ValueError()
```
->
```
 raise ValueError('Currently can only take daystep')
```

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/1/out~testInput:
```diff
@@ -0,0 +1,5828 @@
+
```
This file the log file. Should be removed

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/1/testInput.serpent.seed:
```diff
@@ -0,0 +1 @@
+1523666757
```
This file is read by SERPENT iteself. It is not needed for testing the interface. Please remove it

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+def parseLine(line):
+  """
+  Parse composition line by deleting whitespace and separating the isotope and atomic density.
```
Please indent this line two spaces.

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+def filterTrace(compDict, percentCutoff):
+  """
+  Filters isotopes with less than percentCutoff for easier calculation.
```
Please indent this line two spaces.

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/2/testInput.serpent:
```diff
@@ -0,0 +1,132 @@
+set title "MSBR, no Th,U-233 8.5E-11, after 400 days 7.75E-1"
```
This file is created by RAVEN based on the original file...It should not be here.
Please remove it

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+def bumatRead(bumatFile, percentCutoff):
+  """
+  Reads serpent .bumat output file and stores the composition in a dictionary.
```
Please indent this line two spaces.

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+def searchKeff(resFile):
+  """
+  Searches and returns the mean keff value in the .res file.
```
Please indent this line two spaces.

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/2/out~testInput:
```diff
@@ -0,0 +1,5828 @@
+
+  _                   .-=-.           .-=-.          .-==-.       
```
This file is the logfile (screen output). It is not functional for testing the interface. Please remove it

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+def keffLineParse(keffLine):
+  """
+  Parses through the anaKeff line in .res file.
```
Please indent this line two spaces.

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+def csvRenderDict(csvFilename, dictionary, header):
+  """
+  Renders csv given the dictionary column 1 = key, column 2 = value.
```
Please indent this line two spaces.

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/2/testInput.serpent.seed:
```diff
@@ -0,0 +1 @@
+1523667329
```
This file is read by SERPENT iteself. It is not needed for testing the interface. Please remove it

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+    @ In, csvFilename, string, path of csv file to be created
+    @ In, dictionary, dictionary, dictionary to be rendered into csv file
+    @ In, header, list, list of length 2 of header strings
+    @ Out, bool, bool, True if successful
```
If the return value isn't a named variable, give it the name of the function (`csvRenderDict` in this case).

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/2/testInput.serpent_dep.m:
```diff
@@ -0,0 +1,105 @@
+
+BU = [ 0.00000E+00 6.66600E-02 ];
+
+DAYS = [ 0.00000E+00 1.00000E+01 ];
```
do you read this in the Interface? If not, remove it

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+def readFileIntoList(file):
+  """ 
+  Reads file into list, every line as element.
```
Please indent this line two spaces.

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+      writer.writerow([key, value])
+  return True
+
+def readFileIntoList(file):
```
"file" is a Python2 built-in command, please rename so as not to overwrite the namespace.

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/2/testInput.serpent_res.m:
```diff
@@ -0,0 +1,1120 @@
+
+% Increase counter:
```
do you read this file in the Code Interface? If not, remove it please

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+def findDeptime(inputFile):
+  """
+  Finds the deptime from the input file.
```
Please indent this line two spaces.

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+def makeCsv(csvFilename, inBumatDict, outBumatDict,
+       keffDict, isoList, inputFile):
+  """ 
+  Renders the  csv as filename with the given bumat dict and keff dict.
```
Please indent this line two spaces.

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/testInput.serpent:
```diff
@@ -0,0 +1,132 @@
+set title "MSBR, no Th,U-233 8.5E-11, after 400 days 7.75E-1"
```
This file is created by RAVEN in the perturbation process. It should not be here. Please remove it

PaulTalbot-INL commented on framework/CodeInterfaces/SERPENT/outputParser.py:
```diff
@@ -0,0 +1,210 @@
+
+
+def makeCsv(csvFilename, inBumatDict, outBumatDict,
+       keffDict, isoList, inputFile):
```
this is plenty short to leave on one line (suggested raven standard is 120 characters)

alfoa commented on tests/framework/CodeInterfaceTests/gold/Serpent/SerpentTest1/sampleInput/1/testInput.serpent:
```diff
@@ -0,0 +1,132 @@
+set title "MSBR, no Th,U-233 8.5E-11, after 400 days 7.75E-1"
```
This file is created by RAVEN, it should not be here. please remove it

alfoa commented on tests/framework/CodeInterfaceTests/gold/Serpent/SerpentTest1/sampleInput/2/testInput.serpent:
```diff
@@ -0,0 +1,132 @@
+set title "MSBR, no Th,U-233 8.5E-11, after 400 days 7.75E-1"
+
+include "/projects/sciteam/bahg/projects/raven/framework/CodeInterfaces/SERPENT/aux-input-files/geom_file"
```
This file is created by RAVEN, it should not be here. please remove it

alfoa commented on tests/framework/CodeInterfaceTests/Serpent/SerpentTest1/sampleInput/1/testInput.serpent_res.m:
```diff
@@ -0,0 +1,1120 @@
+
+% Increase counter:
```
do you read this file in the Interface ? if not, please remove it.

PaulTalbot-INL commented:
Additionally, the files
```
framework/CodeInterfaces/SERPENT/SerpentInterface.py
framework/CodeInterfaces/SERPENT/outputParser.py
```
still have some trailing whitespace in them; you can use `raven/developer_tools/delete_trailing_whitespace.sh` on them, or any other tool (most editors can enable automatic removal).

alfoa commented on doc/user_manual/existing_interfaces.tex:
```diff
@@ -2591,3 +2591,138 @@ \subsubsection{DataObjects}
+
+Note that due to the number of isotopes to track,
+we use the \xmlnode{ExternalXML} to import from a file.
+The user must edit the \texttt{aux-input-files} so that it
```
the aux-input-file, should not be committed and added in ```framework/CodeInterfaces/SERPENT```. It is an input and should be explained here and placed in the tests folder

alfoa commented on doc/user_manual/existing_interfaces.tex:
```diff
@@ -2591,3 +2591,138 @@ \subsubsection{DataObjects}
+
+\begin{enumerate}
+\item Edit \texttt{aux-input-files/isoFile} to match isotope nomenclature in SERPENT
+\item run \texttt{python generateCustomVariable.py}
```
same as above

alfoa commented:
@jbae11 we added some comments and request changes.
In addition, you are adding binary files (testInput.serpent.dep) in the tests...do you need them?
Another small request...your test case is huge... (output files > 260,000 lines). Can you shrink the test case to something extremely simple that drastically reduces the size of the outputs? (~1000 lines max)??

PS. you added the test, but no "gold" files have been added

alfoa commented on doc/user_manual/existing_interfaces.tex:
```diff
@@ -2591,3 +2591,138 @@ \subsubsection{DataObjects}
+\texttt{[input].bumat0} and \texttt{[input].bumat1} file to obtain the
+input / output compositions. In the output csv file, the input compositions
+are prepended with the letter `f', and the output compositions with the letter `d'
+(e.g. f92235, d94239), for separation. 
```
Please add a snapshot of the SERPENT outputs here (very small or made up) and the resulting CSV structure (very small as well)

jbae11 commented:
Thank you so much for taking your time to do this.

Regarding binary files:
I was trying to replicate an actual SERPENT run, but I do understand that not all output fiels of SERPENT are needed. Thus, I deleted all output files except the `_res.m`  and `.bumat` files, which are used by the interface to create the csv.

Regarding the gold file:
If I understood it correctly, I added `CodeInterfacesTests/gold/sSerpent/SerpentTest1/RAVENout.csv` which is what the interface should generate in the test run. If this is not the correct way of doing the tests, please let me know.

Regarding the `isoFile`:
as I mentioned in the previous comment, I am not sure if the `isoFile` is the best way for the user to designate isotopes to track. If there is a better way. I'll be happy to change things around!

I have made edits to reflect your comments. Please let me know if there are anything else. Thanks!

alfoa commented:
@jbae11 I will try to come up with a solution this week....

alfoa commented:
@jbae11

alfoa commented:
@jbae11 news?

alfoa commented:
@jbae11 

I changed the destination branch from devel to a branch I created, so I can modify the code on my own.
Let me know if you have anything against this process.

Andrea

alfoa commented:
Since i changed the base branch, I am going to merge this PR in a branch I can work on. I will tag @jbae11  on the new PR once is completed

@PaulTalbot-INL @wangcj05 @mandd @crisrab @joshua-cogliati-inl
